### DISCUSSION TITLE
Validating Accumulated Precipitation Between Forecast Hours

### DISCUSSION THREAD

**Discussion by jh-206** [blaylockbk/Herbie#394] [Q&A]
I am trying to use accumulated precipitation between the f02 and f03 forecast cycles as hourly rain. I am seeing some small discrepancies between the GRIB_stepRange : 2-3 data and what I calculate in the way I assume it is working. The differences are very small, so I'm happy to ignore them, but I want to confirm that I am using the data I think I am using.

Here is a reproducible example. From Herbie objects, I extract the 2-3 hour accumulated and compare to the difference between the 0-3 and 0-2 hour accumulated. The resulting data have some small differences, but I expect zero difference and they are larger than computer error. Are the forecasts changing slightly between the Herbie objects I am requesting?

```
from herbie import Herbie

# Get 3 hour forecast and extract 2-3 and 0-3
H = Herbie(
    "2024-12-17 12:00", 
    model="hrrr", 
    product="prs",
    fxx=3
)
ds = H.xarray(":APCP:surface:2-3 hour acc")
ds3 = H.xarray(":APCP:surface:0-3 hour acc")

# Get 2 hour forecast and extract 0-2
H2 = Herbie(
    "2024-12-17 12:00", 
    model="hrrr", 
    product="prs",
    fxx=2
)
ds2 = H2.xarray(":APCP:surface:0-2 hour acc")

# Calculate 2-3 hour manually
f03 = ds3.tp - ds2.tp

# Calculate Differences
print(np.mean(f03 == ds.tp).values) # about 0.85
print(np.max(np.abs(f03-ds.tp)).values) # 0.00100327
```

Thanks to Brian for great packages!

**Comment by williamhobbs** [comment]
If you suspect that the HRRR output itself is inconsistent, you might ask about it here: https://github.com/NOAA-GSL/GSL-Regional-Model-Forum/discussions. 

And if I'm skimming correctly, you are seeing the max difference at any grid cell being about 0.001 mm of accumulated precipitation. That seems small enough to be some sort of rounding error within HRRR, but maybe I'm misunderstanding.

**Reply by jh-206** [reply]
Thanks for the response. I'm not sure whether it is the HRRR output itself or the way I am accessing the data.

0.001 mm of rain is definitely negligible for a continent-wide computational weather model, but it is large enough where I can't write it off as computer rounding error. So it is signifying to me that I might be using the wrong data or I might not understand how the accumulated rain is being calculated from the model forecasts

**Reply by williamhobbs** [reply]
That makes sense - I agree that unexplainable math errors are not good!

If you haven't seen these, they might be helpful:
- https://herbie.readthedocs.io/en/2024.5.0/user_guide/tutorial/bonus_notebooks/accumulated_precip_ACPC.html
- https://github.com/blaylockbk/Herbie/discussions/10

I haven't worked with precip data, but I have used a similar approach for solar applications, and I think what you are doing is what I would do. I've seen this for average cloud cover: https://solarforecastarbiter-core.readthedocs.io/en/stable/generated/solarforecastarbiter.reference_forecasts.forecast.unmix_intervals.html. And I think I've done something similar for integrated solar insolation from ECMWF.
